File: Week 1/Problem Set 1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

// Prototype
bool is_valid_credit_card(long credit_card);
int get_card_type(long credit_card);

int main(void)
{
    // Prompt the user for a credit card number
    long credit_card = get_long("Credit Card: ");

    // Check if the credit card is valid
    if (is_valid_credit_card(credit_card))
    {
        // Determine and print the card type
        int card_type = get_card_type(credit_card);

        // Print the result based on card type
        switch (card_type)
        {
            case 1:
                printf("AMEX\n");
                break;
            case 2:
                printf("MASTERCARD\n");
                break;
            case 3:
                printf("VISA\n");
                break;
            default:
                printf("INVALID\n");
                break;
        }
    }
    else
    {
        // Print INVALID if the credit card is not valid
        printf("INVALID\n");
    }

    // Return 0 to indicate successful execution
    return 0;
}

// Function to check if a credit card is valid using Luhn's algorithm
bool is_valid_credit_card(long credit_card)
{
    int sum = 0;
    bool is_alternate_digit = false;

    // Iterate through each digit of the credit card number
    while (credit_card > 0)
    {
        // Extract the last digit
        int digit = credit_card % 10;

        // Apply Luhn's algorithm
        if (is_alternate_digit)
        {
            digit *= 2;
            sum += (digit % 10) + (digit / 10);
        }
        else
        {
            sum += digit;
        }

        // Toggle is_alternate_digit for the next iteration
        is_alternate_digit = !is_alternate_digit;

        // Move to the next digit
        credit_card /= 10;
    }

    // The credit card is valid if the sum's last digit is 0
    return (sum % 10 == 0);
}

// Function to determine the card type based on the provided rules
int get_card_type(long credit_card)
{
    // Calculate the number of digits in the credit card number using a loop
    int num_digits = 0;
    long temp = credit_card;

    while (temp != 0)
    {
        temp /= 10;
        num_digits++;
    }

    // Extract the first two digits
    int first_two_digits = credit_card / pow(10, num_digits - 2);

    // Check the rules to determine the card type
    if (num_digits == 15 && (first_two_digits == 34 || first_two_digits == 37))
    {
        return 1; // AMEX
    }
    else if (num_digits == 16 && (first_two_digits >= 51 && first_two_digits <= 55))
    {
        return 2; // MASTERCARD
    }
    else if ((num_digits == 13 || num_digits == 16) && first_two_digits / 10 == 4)
    {
        return 3; // VISA
    }
    else
    {
        return 0; // INVALID
    }
}

File: Week 6/Problem Set 6/sentimental-credit/credit.py
import re
from cs50 import get_string


def main():
    # Get user input for credit card number
    credit_card = get_string("Number: ")

    # Check if the input is a sequence of digits of length 13 to 16
    if re.match(r"^\d{13,16}$", credit_card):
        # Validate the credit card number
        if is_valid_credit_card(credit_card):
            # Determine the card type and print it
            print(get_card_type(credit_card))
        else:
            # If not valid, print "INVALID"
            print("INVALID")
    else:
        # If not a valid sequence of digits, print "INVALID"
        print("INVALID")


def is_valid_credit_card(credit_card):
    """Check if a credit card is valid using Luhn's algorithm."""
    digits = [int(d) for d in str(credit_card)]
    checksum = 0
    for i, digit in enumerate(digits[::-1]):
        if i % 2 == 0:
            checksum += digit
        else:
            checksum += sum(divmod(digit * 2, 10))
    return checksum % 10 == 0


def get_card_type(credit_card):
    """Determine the card type based on the provided rules."""
    num_digits = len(str(credit_card))
    first_two_digits = int(str(credit_card)[:2])
    if num_digits == 15 and first_two_digits in [34, 37]:
        return "AMEX"
    elif num_digits == 16 and 51 <= first_two_digits <= 55:
        return "MASTERCARD"
    elif num_digits in [13, 16] and str(credit_card)[0] == "4":
        return "VISA"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()
